src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,702 @@
+    //      equivalence class and have found that...
+    //
+    //     A. The next permutation is an automorphism of some previous
+    //        permutation of the same equivalence class.
+    //
+    //     B. The next permutation is equivalent to some other existing group
+    //        shape already in a different nontrivial equivalent class.
+    //
+    //     C. The next permutation is equivalent to some other existing group
+    //        shape not yet included in a nontrivial equivalence class.
+    //
+    //   2. We are inserting a new group shape as of yet unaffiliated with a
+    //      nontrival equivalence class and have found that...
+    //
+    //     A. It is the same shape as some group in an existing nontrival
+    //        equivalence class.
+    //
+    //     B. It is the same shape as some other group also not yet affiliated
+    //        with a nontrival equivalence class, so we have a new nontrivial
+    //        equivalence class.
+    //
+    // These possibilities are handled in order below.
+
+    size_t other = it->second;
+
+    auto& groupInfo = groups[group];
+    auto& otherInfo = groups[other];
+
+    size_t groupRep = equivalenceClasses.getRoot(group);
+    size_t otherRep = equivalenceClasses.getRoot(other);
+
+    // Case 1A: There is an automorphism. Skip the rest of the permutations,
+    // which will also be automorphic to previous permutations.
+    if (groupRep == otherRep) {
+      assert(groups[groupRep].classInfo);
+      auto& classInfo = *groups[groupRep].classInfo;
+
+      // Move to the next permutation after advancing the type brand to skip
+      // further automorphisms.
+      classInfo.advanceBrand();
+      classInfo.permute(groupInfo);
+
+      updateShape(group);
+      return;
+    }
+
+    // Case 1B: There is a conflict with a group from a different nontrivial
+    // equivalnece class. Because we canonicalize the shapes whenever we first
+    // create a nontrivial equivalence class, this is usually not possible
+    // because two equivalence classes with isomorphic types should actually be
+    // the same equivalence class. The exception is when the base shapes of two
+    // different equivalence classses are the same, but there is still a
+    // conflict due to brands in each class matching the base type of the other
+    // class. In this case we don't actually want to join the classes; we just
+    // want to advance to the next configuration to resolve the conflict.
+    if (groups[groupRep].classInfo && groups[otherRep].classInfo) {
+      auto& classInfo = *groups[groupRep].classInfo;
+      classInfo.advance();
+      classInfo.permute(groupInfo);
+      updateShape(group);
+      return;
+    }
+
+    // Case 1C: We have permuted ourselves into equivalence with some other
+    // group unaffiliated with a nontrivial equivalence class. Bring that other
+    // group into our equivalence class and try the next permutation instead.
```

Comment:
Why bring that other unaffiliated group into this class, at this time? Is that an optimization, or necessary for correctness?

Reply:
Necessary for correctness, insofar we consider deviations from the intended lazy/eager hybrid design to be correctness bugs. We've discovered that the groups in the current equivalence class are isomorphic to some other group, so logically that group is part of the current equivalence class. Not reflecting that in the data structures means that in the future we might have two "different" equivalence classes that are generating the same shapes as each other and wasting a lot of work.

Reply:
But won't `updateShape` be called on that group later (from the main loop, line 352)? That is, when we get to that group, wouldn't it see the conflict? I may be missing something here.

Reply:
No, this case arises when the unaffiliated group has been processed _before_ the current group.

Reply:
Ah, then what I am confused about is the text

> Bring that other group into our equivalence class and try the next permutation instead

I would expect the reverse. That is, I'd expect that we leave that existing group as it is - it is fine, it's already handled - and we keep modifying ourselves until we are no longer in conflict with anything?

Reply:
Yes, that's what happens. After we join `other` into the equivalence class, it's `group` rather than `other` that is permuted. So we're leaving `other` the way it is, but still bringing it into the equivalence class so that the next time we see a group with the same shape as `other`, we know it should be joined into this equivalence class and reordered to use the next permutation for the class.

Reply:
Isn't `other` already in the equivalence class? You said we've already processed it. Doesn't the result of processing leave the group in the proper equivalence class?

Reply:
No, that would only be the case if we eagerly canonicalized every rec group.

Say we have rec groups `0`, `1`, and `2` with shapes `A` and `A'` that are isomorphic to each other. We might process them in this order:

```
0: A
1: A'
2: A
```
After processing `0` and `1`, we're going to have two different equivalence classes in the data structures because we don't yet know that `A` and `A'` are isomorphic. Then we process `2`, which has the same shape as `0`. We join them into a nontrivial equivalence class, canonicalize their shapes, and reassign the new shapes to `0` and `2`. But what if the new shape we try to give `2` turns out to be `A'`? Only now do we know that the class containing `0` and `2` should be joined with `1`'s class. We update the classes, then try again assigning `2` a new shape `A''`.

Reply:
Given those rec groups and shapes, wouldn't the simpler thing be to notice that `A` and `A'` are isomorphic when we get to `1`? And then fix that up, so that at every point in time, every processed group is unique modulo isomorphism and every group we see later just needs to be modified so it is not isomorphic to them all.

(Sorry, I'm probably missing something basic here!)

Reply:
It would be simpler, but it would also be expensive. The canonicalization routine is something like O(n^2 log n), so we save a lot of work by canonicalizing lazily once we discover a non-trivial equivalence class.

I'll push my commit that describes the high-level algorithmic design better. Maybe that will help.

Reply:
Oh, thanks... now I think I see.

This does seem more complicated than I'd expect. Did you measure canonicalization as being so slow in practice as to require this?

But I guess I'm ok either way. The nice thing in compilers is that passes are modular, so if we end up hitting hard to fix bugs in this pass, we can rewrite it.

Reply:
I hacked up a version that does eager canonicalization and it runs in about the same time, which I guess makes sense because most of the recursion groups are tiny.